### PR TITLE
Restore backward compatibility with GitHub enterprise 2.20

### LINKED ISSUE #16043: Restore campaigns compatibility with GitHub enterprise 2.20
After introducing draft changesets, we no longer support GitHub enterprise <2.21. Depending on how many customers are still on 2.20, we should consider restoring compatibility. 

https://github.com/sourcegraph/sourcegraph/pull/16042 is a draft PR (no pun intended) that would solve it.

### PR DESCRIPTION
**Pull request by eseliger**
This PR restores compatibility with GitHub enterprise 2.20, which seems to be still used by customers. I think once we deprecate support for that version we should revert this PR as it's kind of a mess.
We do so by fetching the instance version first and then constructing the GraphQL queries based on version constraints. 

Any good ideas how to test this besides the manual tests I did? Not sure how much effort we want to put in this PR, given it's out-of-band support work for this sprint.

Closes #16043

### CODE REVIEW

File: cmd/repo-updater/repos/github.go [hunk 1]
```diff
@@ -114,6 +114,17 @@ func newGithubSource(svc *ExternalService, c *schema.GitHubConnection, cf *httpc
...
+	v3Client := github.NewV3Client(apiURL, token, cli)
+	v4Client := github.NewV4Client(apiURL, token, cli)
+
+	if !githubDotCom {
+		enterpriseVersion, err := v3Client.GetEnterpriseVersion(context.Background())
+		if err != nil {
+			return nil, err
+		}
+		v4Client.EnterpriseVersion = enterpriseVersion
+	}
```
Comment (eseliger) [inline cmd/repo-updater/repos/github.go:126]: this would need to be lazy and cached across NewClient calls

File: enterprise/internal/campaigns/state.go [hunk 2]
```diff
@@ -333,6 +333,10 @@ func parseGithubCheckState(s string) campaigns.ChangesetCheckState {
 }
 
 func parseGithubCheckSuiteState(status, conclusion string) campaigns.ChangesetCheckState {
+	// Schema Changes for 2020-02-05
+	// Enum value STALE was added to enum CheckConclusionState
+	// Enum value SKIPPED was added to enum CheckConclusionState
```
Comment (eseliger) [inline enterprise/internal/campaigns/state.go:338]: this probably should be a separate PR but I found it while looking through the docs

File: internal/extsvc/github/pulls.go [hunk 3]
```diff
@@ -508,13 +506,19 @@ type CreatePullRequestInput struct {
 	// The body of the pull request (optional).
 	Body string `json:"body"`
 	// When true the PR will be in draft mode initially.
-	Draft bool `json:"draft"`
+	Draft bool `json:"-"`
```
Comment (mrnugget) [inline internal/extsvc/github/pulls.go:509]: I'd appreciate an addition to the docstring that explains this :)

File: internal/extsvc/github/pulls.go [hunk 4]
```diff
@@ -908,6 +967,22 @@ func (c *V4Client) loadRemainingTimelineItems(ctx context.Context, prID string,
...
+var ghe221PlusOrDotComSemver, _ = semver.NewConstraint(">= 2.21.0")
+
+func timelineItemTypes(version *semver.Version) (string, error) {
+	if ghe220Semver.Check(version) {
+		return timelineItemTypesFmtStr, nil
+	}
+	if ghe221PlusOrDotComSemver.Check(version) {
+		return timelineItemTypesFmtStr + `, CONVERT_TO_DRAFT_EVENT`, nil
+	}
+	return "", errors.New("unsupported version of GitHub")
```
Comment (mrnugget) [inline internal/extsvc/github/pulls.go:983]:
```suggestion
	return "", fmt.Errorf("unsupported version of GitHub: %s", version)
```

File: internal/extsvc/github/version.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	mu        sync.RWMutex
+	versions  map[string]*semver.Version
+	lastReset time.Time
+}
+
+var globalVersionCache *versionCache = &versionCache{
+	versions: make(map[string]*semver.Version),
+}
+
+// normalizeURL will attempt to normalise rawURL.
```
Comment (mrnugget) [inline internal/extsvc/github/version.go:32]: `normali*z*eURL` _and_ `normali*s*e` in the same sentence? You're playing with fire here, buddy

File: internal/extsvc/github/version.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+		return version
+	}
+
+	globalVersionCache.mu.RUnlock()
+	globalVersionCache.mu.Lock()
+	defer globalVersionCache.mu.Unlock()
+	version := c.fetchGitHubVersion(ctx)
+	globalVersionCache.versions[url] = version
+	return version
+}
```
Comment (mrnugget) [inline internal/extsvc/github/version.go:75]: The locking/unlocking in this method scares me a bit, especially since they're used in conditionals etc.

I suggest you try to use small methods, defined in `*versionCache`, to hide the locks.

Something like this:

```go
func (vc *versionCache) VersionForURL(url string) (*semver.Version, bool) {
	vc.mu.RLock()
	defer vc.mu.RUnlock()

	v, ok := vc.versions[url]
	return v, ok
}

func (vc *versionCache) ResetIfExpired() {
	vc.mu.Lock()
	defer vc.mu.Unlock()

	last := vc.lastReset
	expirationTime := last.Add(versionCacheResetTime)

	if last.IsZero() || time.Now().After(expirationTime) {
		vc.lastReset = time.Now()
		vc.versions = make(map[string]*semver.Version)
	}
}

func (vc *versionCache) SetVersion(url string, versioner func() *semver.Version) *semver.Version {
	vc.mu.Lock()
	defer vc.mu.Unlock()

	version := versioner()

	vc.versions[url] = version
	return version
}

// determineGitHubVersion returns a *semver.Version for the targetted GitHub instance by this client. When an
// error occurs, we print a warning to the logs but don't fail and return the allMatchingSemver.
func (c *V4Client) determineGitHubVersion(ctx context.Context) *semver.Version {
	url := normalizeURL(c.apiURL.String())

	globalVersionCache.ResetIfExpired()

	if version, ok := globalVersionCache.VersionForURL(url); ok {
		return version
	}

	return globalVersionCache.SetVersion(url, func() *semver.Version {
		return c.fetchGitHubVersion(ctx)
	})
}
```

**Review by mrnugget — APPROVED** [review]
Thanks for this! Looks good to me, except for the locking/unlocking.

The other thought I had: do we need time-based expiration? I don't think customers update that often and telling them to restart Sourcegraph (or just repo-updater) after upgrading wouldn't be a big deal. Then again, manual steps are always cumbersome so...

File: internal/extsvc/github/pulls.go (hunk 3 above)
Comment (eseliger) [reply internal/extsvc/github/pulls.go:509]: Whoups that can be reverted.

File: internal/extsvc/github/version.go (hunk 5 above)
Comment (eseliger) [reply internal/extsvc/github/version.go:32]: That happens when you copy 🏴󠁧󠁢󠁥󠁮󠁧󠁿 code :)

File: internal/extsvc/github/version.go (hunk 6 above)
Comment (eseliger) [reply internal/extsvc/github/version.go:75]: To be honest, I think the RWMutex was a bit overkill, really 😬 I'll just use a simple lock on the function.

File: internal/extsvc/github/version.go (hunk 5 above)
Comment (LawnGnome) [reply internal/extsvc/github/version.go]: :australia: :canada:
